README.md
# rucnv
Encoding converter for texts in Russian language

## General Information
Up to 14 different encodings are used to represent texts in Russian language [(koi-7, koi-8r, cp866, cp1251, cp10007, utf-8, ...)](http://litwr2.atspace.eu/ru_sheet1.pdf)!  This program can perform conversion between them. It also can produce Unicode names for the symbol codes.

It shows the unicodes that cannot be converted.  It allows to put two types of substitutions for the inconvertible signs.  It may be a unicode number or a similar sign.

This program may also be used for transliteration (reversible substitution Cyrillic letters by Latin ones).  Transliteration is the only way to make deal with Russian texts on the system without Cyrillic fonts.  RUCNV knows two types of transliteration: GOST 7.79-2000 and KOI-7 based.

RUCNV may be used for conversion any Russian or Unicode texts into the form which will be accepted by any 8-bit text editor and for followed backward conversion. 

This program is distributed under the GNU General Public License.


## Usage
	USAGE: rucnv <translation> <source> <destination>
	       rucnv -V       output version information and exit
	       rucnv --help   display this help and exit
	<translation> has form <source coding><result coding>{s}{u}
	  's' ('u') makes (unicode) substitutions for missing characters
	codings:
	  0: Unicode (UCS-2 Little Endian)	  1: UTF-8
	  2: Koi8-r				  3: CP866 (MS-DOS)
	  4: ISO 8859-5				  5: CP1251 (Microsoft Windows)
	  6: Macintosh Cyrillic (CP10007)	  7: Modified Alternative (MS-DOS)
	  8: Alternative (MS-DOS)		  9: Main (MS-DOS)
	  A: Bulgarian, MIC, Interprog (MS-DOS)	  B: Koi-8
	  C: Koi-7				  D: Transliteration (Koi-7 based)
	  E: Transliteration (GOST 7.79-2000)	  F: Unicode Names
	  G: UTF-16BE
	Examples:
	  rucnv 52s wintext.txt lintext.txt
	  rucnv 31 dostext.txt utf8text.txt
	  rucnv 12u utf8text.txt koi8rtext.txt
	  rucnv 50u wintext.txt ucs2text.txt


## Basic Installation
You need C++ compiler and C++ Standard Library.

The simplest way to make this program is:

1. Edit makefile to set correct MANDIR and BINDIR for your system.
2. Type `make' to compile the program.
3. Type `make install' to install the program and documentation.


## Microsoft DOS/Windows Installation
The simplest way to make this with DELORIE DJGPP V2 system for C/C++ 
(http://www.delorie.com) program is:

1. Type `gxx -o rucnv.exe rucnv.cpp -lstdcxx` to get the executable file.
2. Type `strip rucnv.exe` to remove symbols from file.
3. Type `exe2coff rucnv.exe` and then 
     `copy /b [dir for djgpp]\bin\cwsdstub.exe + rucnv rucnv.exe` to get the 
     executable file independent from DJGPP system.

Steps 2 and 3 are optional.


## A NOTE ON USAGE
You may use rucnv with the same filenames for input and output.


## BUGS, SUGGESTIONS, ETC.
Please write me if you have trouble running rucnv, or if you 
have suggestions or patches.

	Vladimir Lidovski
	litwr@yandex.ru
	http://litwr2.atspace.eu


## COPYRIGHT/LICENSE
All source code is Copyright (C) 2002 Lidovski Vladimir

This program is distributed under the GNU General Public License, Version
2, or, at your discretion, any later version. The GNU General Public License
is available via the Web at <http://www.gnu.org/copyleft/gpl.html>. The GPL
is designed to allow you to alter and redistribute the package, as long as
you do not remove that freedom from others.


NEWS
version 1.07:  The C++ syntax is adjusted to the new C++ standards
version 1.08:  Added UTF-16BE support. Fixed minor compilation problem


subtab.h
struct SubTab {
  int uc;
  const char *sn;
} subtab[] = {{0xA0," "},{0x00A1,"!"},{0x00A3,"<pound>"},{0x00A4,"o"},
{0x00A6,"|"},
{0x00A7,"S"},{0x00A9,"(c)"},{0x00AA,"_a"},{0x00AB,"<<"},{0x00AC,"<not>"},
{0x00AD,"-"},{0x00AE,"(R)"},{0x00B0,"o"},{0x00B1,"+-"},{0x00B2,"^2"},
{0x00B5,"<micro>"},{0x00B6,"\\|"},{0x00B7,"."},{0x00BA,"_o"},{0x00BB,">>"},
{0x00BC,"1/4"},{0x00BD,"1/2"},{0x00BF,"?"},{0x00D1,"~N"},{0x00DF,"ss"},
{0x00E1,"'a"},{0x00ED,"'i"},{0x00F1,"~n"},{0x00F3,"'o"},{0x00F7,"/"},
{0x00FA,"'u"},{0x0192,"f"},{0x0393,"<Gamma>"},{0x0398,"<Theta>"},
{0x03A3,"<Sigma>"},{0x03A6,"<Phi>"},{0x03A9,"<Omega>"},{0x03B1,"<alpha>"},
{0x03B4,"<delta>"},{0x03B5,"<epsilon>"},{0x03C0,"<pi>"},{0x03C3,"<sigma>"},
{0x03C4,"<tau>"},{0x03C6,"<phi>"},{0x0401,"\"E"},{0x0402,"<Dje>"},
{0x0403,"<Gje>"},{0x0404,"<E>"},{0x0405,"S"},{0x0406,"I"},{0x0407,"\"I"},
{0x0408,"J"},{0x0409,"<Lje>"},{0x040A,"<Nje>"},{0x040B,"<Tshe>"},
{0x040C,"'K"},{0x040E,"~Y"},{0x040F,"<Dzhe>"},{0x042A,"<Hard>"},
{0x0451,"\"e"},
{0x0452,"<dje>"},{0x0453,"<gje>"},{0x0454,"<e>"},{0x0455,"s"},
{0x0456,"i"},{0x0457,"\"i"},{0x0458,"j"},{0x0459,"<lje>"},{0x045A,"<nje>"},
{0x045B,"<tshe>"},{0x045C,"'k"},{0x045E,"~y"},{0x045F,"<dzhe>"},
{0x0490,"<Ghe upturn>"},{0x0491,"<ghe upturn>"},{0x2013,"-"},{0x2014,"--"},
{0x2018,"`"},{0x2019,"'"},{0x201A,","},{0x201C,"``"},{0x201D,"''"},
{0x201E,",,"},{0x2020,"<dagger>"},{0x2021,"<dbl dagger>"},{0x2022,"*"},
{0x2026,"..."},{0x2030,"%o"},{0x2039,"<"},{0x203A,">"},{0x207F,"^n"},
{0x20AC,"<Euro>"},{0x2116,"No"},{0x2122,"TM"},{0x2190,"<-"},{0x2191,"^"},
{0x2192,"->"},{0x2193,"v"},{0x2206,"<increment>"},{0x2219,"."},
{0x221A,"<sqrt>"},{0x221E,"<infty>"},{0x2229,"<cap>"},{0x2248,"<approx>"},
{0x2260,"/="},{0x2261,"<identical>"},{0x2264,"<="},{0x2265,">="},
{0x2310,"<reversed not>"},{0x2320,"<top int>"},{0x2321,"<bottom int>"},
{0x2500,"-"},{0x2502,"|"},{0x250c,"."},{0x2510,"."},{0x2514,"`"},
{0x2518,"'"},{0x251c,"}"},{0x2524,"{"},{0x252c,"+"},{0x2534,"`"},
{0x253c,"+"},{0x2550,"="},{0x2551,"I"},{0x2552,"."},{0x2553,"."},
{0x2554,"."},{0x2555,"."},{0x2556,"."},{0x2557,"."},{0x2558,"`"},
{0x2559,"`"},{0x255a,"`"},{0x255b,"'"},{0x255c,"'"},{0x255d,"'"},
{0x255e,"}"},{0x255f,"}"},{0x2560,"}"},{0x2561,"{"},{0x2562,"{"},
{0x2563,"{"},{0x2564,"*"},{0x2565,"+"},{0x2566,"*"},{0x2567,"*"},
{0x2568,"+"},{0x2569,"*"},{0x256a,"*"},{0x256b,"*"},{0x256c,"*"},
{0x2580,"~"},{0x2584,"_"},{0x2588,"#"},{0x258c,"]"},{0x2590,"["},
{0x2591,":"},{0x2592,"%"},{0x2593,"&"},{0x25a0,"="},
{0xe000,"/"}, {0xe001,"\\"},{0xe002,"/"},{0xe003,"\\"}};


uninames.h
char unidata[] = "NULL\0START OF HEADING\0START OF TEXT\0END OF TEXT\0END OF TRANSMISSION\0\
ENQUIRY\0ACKNOWLEDGE\0BELL\0BACKSPACE\0HORIZONTAL TABULATION\0LINE FEED\0\
VERTICAL TABULATION\0FORM FEED\0CARRIAGE RETURN\0SHIFT OUT\0SHIFT IN\0\
DATA LINK ESCAPE\0DEVICE CONTROL ONE\0DEVICE CONTROL TWO\0\
DEVICE CONTROL THREE\0DEVICE CONTROL FOUR\0NEGATIVE ACKNOWLEDGE\0\
SYNCHRONOUS IDLE\0END OF TRANSMISSION BLOCK\0CANCEL\0END OF MEDIUM\0SUBSTITUTE\0\
ESCAPE\0FILE SEPARATOR\0GROUP SEPARATOR\0RECORD SEPARATOR\0UNIT SEPARATOR\0\
SPACE\0EXCLAMATION MARK\0QUOTATION MARK\0NUMBER SIGN\0DOLLAR SIGN\0PERCENT SIGN\0\
AMPERSAND\0APOSTROPHE\0LEFT PARENTHESIS\0RIGHT PARENTHESIS\0ASTERISK\0PLUS SIGN\0\
COMMA\0HYPHEN-MINUS\0FULL STOP\0SOLIDUS\0DIGIT ZERO\0DIGIT ONE\0DIGIT TWO\0\
DIGIT THREE\0DIGIT FOUR\0DIGIT FIVE\0DIGIT SIX\0DIGIT SEVEN\0DIGIT EIGHT\0\
DIGIT NINE\0COLON\0SEMICOLON\0LESS-THAN SIGN\0EQUALS SIGN\0GREATER-THAN SIGN\0\
QUESTION MARK\0COMMERCIAL AT\0LATIN CAPITAL LETTER A\0LATIN CAPITAL LETTER B\0\
LATIN CAPITAL LETTER C\0LATIN CAPITAL LETTER D\0LATIN CAPITAL LETTER E\0\
LATIN CAPITAL LETTER F\0LATIN CAPITAL LETTER G\0LATIN CAPITAL LETTER H\0\
LATIN CAPITAL LETTER I\0LATIN CAPITAL LETTER J\0LATIN CAPITAL LETTER K\0\
LATIN CAPITAL LETTER L\0LATIN CAPITAL LETTER M\0LATIN CAPITAL LETTER N\0\
LATIN CAPITAL LETTER O\0LATIN CAPITAL LETTER P\0LATIN CAPITAL LETTER Q\0\
LATIN CAPITAL LETTER R\0LATIN CAPITAL LETTER S\0LATIN CAPITAL LETTER T\0\
LATIN CAPITAL LETTER U\0LATIN CAPITAL LETTER V\0LATIN CAPITAL LETTER W\0\
LATIN CAPITAL LETTER X\0LATIN CAPITAL LETTER Y\0LATIN CAPITAL LETTER Z\0\
LEFT SQUARE BRACKET\0REVERSE SOLIDUS\0RIGHT SQUARE BRACKET\0CIRCUMFLEX ACCENT\0\
LOW LINE\0GRAVE ACCENT\0LATIN SMALL LETTER A\0LATIN SMALL LETTER B\0\
LATIN SMALL LETTER C\0LATIN SMALL LETTER D\0LATIN SMALL LETTER E\0\
LATIN SMALL LETTER F\0LATIN SMALL LETTER G\0LATIN SMALL LETTER H\0\
LATIN SMALL LETTER I\0LATIN SMALL LETTER J\0LATIN SMALL LETTER K\0\
LATIN SMALL LETTER L\0LATIN SMALL LETTER M\0LATIN SMALL LETTER N\0\
LATIN SMALL LETTER O\0LATIN SMALL LETTER P\0LATIN SMALL LETTER Q\0\
LATIN SMALL LETTER R\0LATIN SMALL LETTER S\0LATIN SMALL LETTER T\0\
LATIN SMALL LETTER U\0LATIN SMALL LETTER V\0LATIN SMALL LETTER W\0\
LATIN SMALL LETTER X\0LATIN SMALL LETTER Y\0LATIN SMALL LETTER Z\0\
LEFT CURLY BRACKET\0VERTICAL LINE\0RIGHT CURLY BRACKET\0TILDE\0DELETE\0\
CYRILLIC CAPITAL LETTER DJE\0CYRILLIC CAPITAL LETTER GJE\0\
SINGLE LOW-9 QUOTATION MARK\0CYRILLIC SMALL LETTER GJE\0\
DOUBLE LOW-9 QUOTATION MARK\0HORIZONTAL ELLIPSIS\0DAGGER\0DOUBLE DAGGER\0\
EURO SIGN\0PER MILLE SIGN\0CYRILLIC CAPITAL LETTER LJE\0\
SINGLE LEFT-POINTING ANGLE QUOTATION MARK\0CYRILLIC CAPITAL LETTER NJE\0\
CYRILLIC CAPITAL LETTER KJE\0CYRILLIC CAPITAL LETTER TSHE\0\
CYRILLIC CAPITAL LETTER DZHE\0CYRILLIC SMALL LETTER DJE\0\
LEFT SINGLE QUOTATION MARK\0RIGHT SINGLE QUOTATION MARK\0\
LEFT DOUBLE QUOTATION MARK\0RIGHT DOUBLE QUOTATION MARK\0BULLET\0EN DASH\0\
EM DASH\0TRADE MARK SIGN\0CYRILLIC SMALL LETTER LJE\0\
SINGLE RIGHT-POINTING ANGLE QUOTATION MARK\0CYRILLIC SMALL LETTER NJE\0\
CYRILLIC SMALL LETTER KJE\0CYRILLIC SMALL LETTER TSHE\0\
CYRILLIC SMALL LETTER DZHE\0NO-BREAK SPACE\0CYRILLIC CAPITAL LETTER SHORT U\0\
CYRILLIC SMALL LETTER SHORT U\0CYRILLIC CAPITAL LETTER JE\0CURRENCY SIGN\0\
CYRILLIC CAPITAL LETTER GHE WITH UPTURN\0BROKEN BAR\0SECTION SIGN\0\
CYRILLIC CAPITAL LETTER IO\0COPYRIGHT SIGN\0\
CYRILLIC CAPITAL LETTER UKRAINIAN IE\0\
LEFT-POINTING DOUBLE ANGLE QUOTATION MARK\0NOT SIGN\0SOFT HYPHEN\0\
REGISTERED SIGN\0CYRILLIC CAPITAL LETTER YI\0DEGREE SIGN\0PLUS-MINUS SIGN\0\
CYRILLIC CAPITAL LETTER BYELORUSSIAN-UKRAINIAN I\0\
CYRILLIC SMALL LETTER BYELORUSSIAN-UKRAINIAN I\0\
CYRILLIC SMALL LETTER GHE WITH UPTURN\0MICRO SIGN\0PILCROW SIGN\0MIDDLE DOT\0\
CYRILLIC SMALL LETTER IO\0NUMERO SIGN\0CYRILLIC SMALL LETTER UKRAINIAN IE\0\
RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK\0CYRILLIC SMALL LETTER JE\0\
CYRILLIC CAPITAL LETTER DZE\0CYRILLIC SMALL LETTER DZE\0\
CYRILLIC SMALL LETTER YI\0CYRILLIC CAPITAL LETTER A\0\
CYRILLIC CAPITAL LETTER BE\0CYRILLIC CAPITAL LETTER VE\0\
CYRILLIC CAPITAL LETTER GHE\0CYRILLIC CAPITAL LETTER DE\0\
CYRILLIC CAPITAL LETTER IE\0CYRILLIC CAPITAL LETTER ZHE\0\
CYRILLIC CAPITAL LETTER ZE\0CYRILLIC CAPITAL LETTER I\0\
CYRILLIC CAPITAL LETTER SHORT I\0CYRILLIC CAPITAL LETTER KA\0\
CYRILLIC CAPITAL LETTER EL\0CYRILLIC CAPITAL LETTER EM\0\
CYRILLIC CAPITAL LETTER EN\0CYRILLIC CAPITAL LETTER O\0\
CYRILLIC CAPITAL LETTER PE\0CYRILLIC CAPITAL LETTER ER\0\
CYRILLIC CAPITAL LETTER ES\0CYRILLIC CAPITAL LETTER TE\0\
CYRILLIC CAPITAL LETTER U\0CYRILLIC CAPITAL LETTER EF\0\
CYRILLIC CAPITAL LETTER HA\0CYRILLIC CAPITAL LETTER TSE\0\
CYRILLIC CAPITAL LETTER CHE\0CYRILLIC CAPITAL LETTER SHA\0\
CYRILLIC CAPITAL LETTER SHCHA\0CYRILLIC CAPITAL LETTER HARD SIGN\0\
CYRILLIC CAPITAL LETTER YERU\0CYRILLIC CAPITAL LETTER SOFT SIGN\0\
CYRILLIC CAPITAL LETTER E\0CYRILLIC CAPITAL LETTER YU\0\
CYRILLIC CAPITAL LETTER YA\0CYRILLIC SMALL LETTER A\0\
CYRILLIC SMALL LETTER BE\0CYRILLIC SMALL LETTER VE\0\
CYRILLIC SMALL LETTER GHE\0CYRILLIC SMALL LETTER DE\0\
CYRILLIC SMALL LETTER IE\0CYRILLIC SMALL LETTER ZHE\0\
CYRILLIC SMALL LETTER ZE\0CYRILLIC SMALL LETTER I\0\
CYRILLIC SMALL LETTER SHORT I\0CYRILLIC SMALL LETTER KA\0\
CYRILLIC SMALL LETTER EL\0CYRILLIC SMALL LETTER EM\0CYRILLIC SMALL LETTER EN\0\
CYRILLIC SMALL LETTER O\0CYRILLIC SMALL LETTER PE\0CYRILLIC SMALL LETTER ER\0\
CYRILLIC SMALL LETTER ES\0CYRILLIC SMALL LETTER TE\0CYRILLIC SMALL LETTER U\0\
CYRILLIC SMALL LETTER EF\0CYRILLIC SMALL LETTER HA\0\
CYRILLIC SMALL LETTER TSE\0CYRILLIC SMALL LETTER CHE\0\
CYRILLIC SMALL LETTER SHA\0CYRILLIC SMALL LETTER SHCHA\0\
CYRILLIC SMALL LETTER HARD SIGN\0CYRILLIC SMALL LETTER YERU\0\
CYRILLIC SMALL LETTER SOFT SIGN\0CYRILLIC SMALL LETTER E\0\
CYRILLIC SMALL LETTER YU\0CYRILLIC SMALL LETTER YA\0POUND SIGN\0NOT EQUAL TO\0\
INFINITY\0LESS-THAN OR EQUAL TO\0GREATER-THAN OR EQUAL TO\0SQUARE ROOT\0\
LATIN SMALL LETTER F WITH HOOK\0ALMOST EQUAL TO\0INCREMENT\0DIVISION SIGN\0\
LIGHT SHADE\0MEDIUM SHADE\0DARK SHADE\0BOX DRAWINGS LIGHT VERTICAL\0\
BOX DRAWINGS LIGHT VERTICAL AND LEFT\0\
BOX DRAWINGS VERTICAL SINGLE AND LEFT DOUBLE\0\
BOX DRAWINGS VERTICAL DOUBLE AND LEFT SINGLE\0\
BOX DRAWINGS DOWN DOUBLE AND LEFT SINGLE\0\
BOX DRAWINGS DOWN SINGLE AND LEFT DOUBLE\0\
BOX DRAWINGS DOUBLE VERTICAL AND LEFT\0BOX DRAWINGS DOUBLE VERTICAL\0\
BOX DRAWINGS DOUBLE DOWN AND LEFT\0BOX DRAWINGS DOUBLE UP AND LEFT\0\
BOX DRAWINGS UP DOUBLE AND LEFT SINGLE\0\
BOX DRAWINGS UP SINGLE AND LEFT DOUBLE\0BOX DRAWINGS LIGHT DOWN AND LEFT\0\
BOX DRAWINGS LIGHT UP AND RIGHT\0BOX DRAWINGS LIGHT UP AND HORIZONTAL\0\
BOX DRAWINGS LIGHT DOWN AND HORIZONTAL\0\
BOX DRAWINGS LIGHT VERTICAL AND RIGHT\0BOX DRAWINGS LIGHT HORIZONTAL\0\
BOX DRAWINGS LIGHT VERTICAL AND HORIZONTAL\0\
BOX DRAWINGS VERTICAL SINGLE AND RIGHT DOUBLE\0\
BOX DRAWINGS VERTICAL DOUBLE AND RIGHT SINGLE\0\
BOX DRAWINGS DOUBLE UP AND RIGHT\0BOX DRAWINGS DOUBLE DOWN AND RIGHT\0\
BOX DRAWINGS DOUBLE UP AND HORIZONTAL\0\
BOX DRAWINGS DOUBLE DOWN AND HORIZONTAL\0\
BOX DRAWINGS DOUBLE VERTICAL AND RIGHT\0BOX DRAWINGS DOUBLE HORIZONTAL\0\
BOX DRAWINGS DOUBLE VERTICAL AND HORIZONTAL\0\
BOX DRAWINGS UP SINGLE AND HORIZONTAL DOUBLE\0\
BOX DRAWINGS UP DOUBLE AND HORIZONTAL SINGLE\0\
BOX DRAWINGS DOWN SINGLE AND HORIZONTAL DOUBLE\0\
BOX DRAWINGS DOWN DOUBLE AND HORIZONTAL SINGLE\0\
BOX DRAWINGS UP DOUBLE AND RIGHT SINGLE\0\
BOX DRAWINGS UP SINGLE AND RIGHT DOUBLE\0\
BOX DRAWINGS DOWN SINGLE AND RIGHT DOUBLE\0\
BOX DRAWINGS DOWN DOUBLE AND RIGHT SINGLE\0\
BOX DRAWINGS VERTICAL DOUBLE AND HORIZONTAL SINGLE\0\
BOX DRAWINGS VERTICAL SINGLE AND HORIZONTAL DOUBLE\0\
BOX DRAWINGS LIGHT UP AND LEFT\0BOX DRAWINGS LIGHT DOWN AND RIGHT\0\
FULL BLOCK\0LOWER HALF BLOCK\0LEFT HALF BLOCK\0RIGHT HALF BLOCK\0\
UPPER HALF BLOCK\0BULLET OPERATOR\0BLACK SQUARE\0TOP HALF INTEGRAL\0\
BOTTOM HALF INTEGRAL\0SUPERSCRIPT TWO\0INVERTED EXCLAMATION MARK\0\
FEMININE ORDINAL INDICATOR\0MASCULINE ORDINAL INDICATOR\0\
VULGAR FRACTION ONE QUARTER\0VULGAR FRACTION ONE HALF\0\
INVERTED QUESTION MARK\0LATIN CAPITAL LETTER N WITH TILDE\0\
LATIN SMALL LETTER SHARP S\0LATIN SMALL LETTER A WITH ACUTE\0\
LATIN SMALL LETTER I WITH ACUTE\0LATIN SMALL LETTER N WITH TILDE\0\
LATIN SMALL LETTER O WITH ACUTE\0LATIN SMALL LETTER U WITH ACUTE\0\
GREEK CAPITAL LETTER GAMMA\0GREEK CAPITAL LETTER THETA\0\
GREEK CAPITAL LETTER SIGMA\0GREEK CAPITAL LETTER PHI\0\
GREEK CAPITAL LETTER OMEGA\0GREEK SMALL LETTER ALPHA\0\
GREEK SMALL LETTER DELTA\0GREEK SMALL LETTER EPSILON\0GREEK SMALL LETTER PI\0\
GREEK SMALL LETTER SIGMA\0GREEK SMALL LETTER TAU\0GREEK SMALL LETTER PHI\0\
SUPERSCRIPT LATIN SMALL LETTER N\0LEFTWARDS ARROW\0UPWARDS ARROW\0\
RIGHTWARDS ARROW\0DOWNWARDS ARROW\0INTERSECTION\0IDENTICAL TO\0\
REVERSED NOT SIGN\0?BOX DRAWINGS LOWER RIGHT CORNER DIAGONAL LINE\0\
?BOX DRAWINGS LOWER LEFT CORNER DIAGONAL LINE\0\
?BOX DRAWINGS UPPER LEFT CORNER DIAGONAL LINE\0\
?BOX DRAWINGS UPPER RIGHT CORNER DIAGONAL LINE\0";

int uniindex[] = {0, 5, 22, 36, 48, 68, 76, 88, 93, 103, 125, 135, 155, 165, 
  181, 191, 200, 217, 236, 255, 276, 296, 317, 334, 360, 367, 381, 392, 399, 
  414, 430, 447, 462, 468, 485, 500, 512, 524, 537, 547, 558, 575, 593, 602, 
  612, 618, 631, 641, 649, 660, 670, 680, 692, 703, 714, 724, 736, 748, 759, 
  765, 775, 790, 802, 820, 834, 848, 871, 894, 917, 940, 963, 986, 1009, 
  1032, 1055, 1078, 1101, 1124, 1147, 1170, 1193, 1216, 1239, 1262, 1285, 
  1308, 1331, 1354, 1377, 1400, 1423, 1446, 1466, 1482, 1503, 1521, 1530, 
  1543, 1564, 1585, 1606, 1627, 1648, 1669, 1690, 1711, 1732, 1753, 1774, 
  1795, 1816, 1837, 1858, 1879, 1900, 1921, 1942, 1963, 1984, 2005, 2026, 
  2047, 2068, 2089, 2108, 2122, 2142, 2148, 2155, 2183, 2211, 2239, 2265, 
  2293, 2313, 2320, 2334, 2344, 2359, 2387, 2429, 2457, 2485, 2514, 2543, 
  2569, 2596, 2624, 2651, 2679, 2686, 2694, 2702, 2718, 2744, 2787, 2813, 
  2839, 2866, 2893, 2908, 2940, 2970, 2997, 3011, 3051, 3062, 3075, 3102, 
  3117, 3154, 3196, 3205, 3217, 3233, 3260, 3272, 3288, 3337, 3384, 3422, 
  3433, 3446, 3457, 3482, 3494, 3529, 3572, 3597, 3625, 3651, 3676, 3702, 
  3729, 3756, 3784, 3811, 3838, 3866, 3893, 3919, 3951, 3978, 4005, 4032, 
  4059, 4085, 4112, 4139, 4166, 4193, 4219, 4246, 4273, 4301, 4329, 4357, 
  4387, 4421, 4450, 4484, 4510, 4537, 4564, 4588, 4613, 4638, 4664, 4689, 
  4714, 4740, 4765, 4789, 4819, 4844, 4869, 4894, 4919, 4943, 4968, 4993, 
  5018, 5043, 5067, 5092, 5117, 5143, 5169, 5195, 5223, 5255, 5282, 5314, 
  5338, 5363, 5388, 5399, 5412, 5421, 5443, 5468, 5480, 5511, 5527, 5537, 
  5551, 5563, 5576, 5587, 5615, 5652, 5697, 5742, 5783, 5824, 5862, 5891, 
  5925, 5957, 5996, 6035, 6068, 6100, 6137, 6176, 6214, 6244, 6287, 6333, 
  6379, 6412, 6447, 6485, 6525, 6564, 6595, 6639, 6684, 6729, 6776, 6823, 
  6863, 6903, 6945, 6987, 7038, 7089, 7120, 7154, 7165, 7182, 7198, 7215, 
  7232, 7248, 7261, 7279, 7300, 7316, 7342, 7369, 7397, 7425, 7450, 7473, 
  7507, 7534, 7566, 7598, 7630, 7662, 7694, 7721, 7748, 7775, 7800, 7827, 
  7852, 7877, 7904, 7926, 7951, 7974, 7997, 8030, 8046, 8060, 8077, 8093, 
  8106, 8119, 8137, 8184, 8230, 8276};

int uninum[] = {0x0000, 0x0001, 0x0002, 0x0003, 0x0004, 0x0005, 0x0006, 
  0x0007, 0x0008, 0x0009, 0x000A, 0x000B, 0x000C, 0x000D, 0x000E, 0x000F, 
  0x0010, 0x0011, 0x0012, 0x0013, 0x0014, 0x0015, 0x0016, 0x0017, 0x0018, 
  0x0019, 0x001A, 0x001B, 0x001C, 0x001D, 0x001E, 0x001F, 0x0020, 0x0021, 
  0x0022, 0x0023, 0x0024, 0x0025, 0x0026, 0x0027, 0x0028, 0x0029, 0x002A, 
  0x002B, 0x002C, 0x002D, 0x002E, 0x002F, 0x0030, 0x0031, 0x0032, 0x0033, 
  0x0034, 0x0035, 0x0036, 0x0037, 0x0038, 0x0039, 0x003A, 0x003B, 0x003C, 
  0x003D, 0x003E, 0x003F, 0x0040, 0x0041, 0x0042, 0x0043, 0x0044, 0x0045, 
  0x0046, 0x0047, 0x0048, 0x0049, 0x004A, 0x004B, 0x004C, 0x004D, 0x004E, 
  0x004F, 0x0050, 0x0051, 0x0052, 0x0053, 0x0054, 0x0055, 0x0056, 0x0057, 
  0x0058, 0x0059, 0x005A, 0x005B, 0x005C, 0x005D, 0x005E, 0x005F, 0x0060, 
  0x0061, 0x0062, 0x0063, 0x0064, 0x0065, 0x0066, 0x0067, 0x0068, 0x0069, 
  0x006A, 0x006B, 0x006C, 0x006D, 0x006E, 0x006F, 0x0070, 0x0071, 0x0072, 
  0x0073, 0x0074, 0x0075, 0x0076, 0x0077, 0x0078, 0x0079, 0x007A, 0x007B, 
  0x007C, 0x007D, 0x007E, 0x007F, 0x0402, 0x0403, 0x201A, 0x0453, 0x201E, 
  0x2026, 0x2020, 0x2021, 0x20AC, 0x2030, 0x0409, 0x2039, 0x040A, 0x040C, 
  0x040B, 0x040F, 0x0452, 0x2018, 0x2019, 0x201C, 0x201D, 0x2022, 0x2013, 
  0x2014, 0x2122, 0x0459, 0x203A, 0x045A, 0x045C, 0x045B, 0x045F, 0x00A0, 
  0x040E, 0x045E, 0x0408, 0x00A4, 0x0490, 0x00A6, 0x00A7, 0x0401, 0x00A9, 
  0x0404, 0x00AB, 0x00AC, 0x00AD, 0x00AE, 0x0407, 0x00B0, 0x00B1, 0x0406, 
  0x0456, 0x0491, 0x00B5, 0x00B6, 0x00B7, 0x0451, 0x2116, 0x0454, 0x00BB, 
  0x0458, 0x0405, 0x0455, 0x0457, 0x0410, 0x0411, 0x0412, 0x0413, 0x0414, 
  0x0415, 0x0416, 0x0417, 0x0418, 0x0419, 0x041A, 0x041B, 0x041C, 0x041D, 
  0x041E, 0x041F, 0x0420, 0x0421, 0x0422, 0x0423, 0x0424, 0x0425, 0x0426, 
  0x0427, 0x0428, 0x0429, 0x042A, 0x042B, 0x042C, 0x042D, 0x042E, 0x042F, 
  0x0430, 0x0431, 0x0432, 0x0433, 0x0434, 0x0435, 0x0436, 0x0437, 0x0438, 
  0x0439, 0x043A, 0x043B, 0x043C, 0x043D, 0x043E, 0x043F, 0x0440, 0x0441, 
  0x0442, 0x0443, 0x0444, 0x0445, 0x0446, 0x0447, 0x0448, 0x0449, 0x044A, 
  0x044B, 0x044C, 0x044D, 0x044E, 0x044F, 0x00A3, 0x2260, 0x221E, 0x2264, 
  0x2265, 0x221A, 0x0192, 0x2248, 0x2206, 0x00F7, 0x2591, 0x2592, 0x2593, 
  0x2502, 0x2524, 0x2561, 0x2562, 0x2556, 0x2555, 0x2563, 0x2551, 0x2557, 
  0x255d, 0x255c, 0x255b, 0x2510, 0x2514, 0x2534, 0x252c, 0x251c, 0x2500, 
  0x253c, 0x255e, 0x255f, 0x255a, 0x2554, 0x2569, 0x2566, 0x2560, 0x2550, 
  0x256c, 0x2567, 0x2568, 0x2564, 0x2565, 0x2559, 0x2558, 0x2552, 0x2553, 
  0x256b, 0x256a, 0x2518, 0x250c, 0x2588, 0x2584, 0x258c, 0x2590, 0x2580, 
  0x2219, 0x25a0, 0x2320, 0x2321, 0x00B2, 0x00A1, 0x00AA, 0x00BA, 0x00BC, 
  0x00BD, 0x00BF, 0x00D1, 0x00DF, 0x00E1, 0x00ED, 0x00F1, 0x00F3, 0x00FA, 
  0x0393, 0x0398, 0x03A3, 0x03A6, 0x03A9, 0x03B1, 0x03B4, 0x03B5, 0x03C0, 
  0x03C3, 0x03C4, 0x03C6, 0x207F, 0x2190, 0x2191, 0x2192, 0x2193, 0x2229, 
  0x2261, 0x2310, 0xe000, 0xe001, 0xe002, 0xe003};
